File: 328.odd-even-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
typedef struct ListNode Node;

/* Approach: Two Pointer, Complexity: O(n), O(1) */
Node* oddEvenList(Node* head){
    /* For empty list, single node list and double node list, input is already
     * equal to result.
     */
    if (!head || !head->next || !head->next->next)
        return head;

    /* Start 'odd' from the first odd node and 'even' from first even node. */
    Node *evenhead, *even, *odd;
    odd = head;
    even = evenhead = head->next;

    /* Link all odd nodes (in order) together. Do the same for even nodes. The
     * even nodes always run out first. So for termination, check if 'even'
     * points to NULL or last node in the list i.e. no more nodes left to link.
     */
    while (even && even->next) {
        odd->next = odd->next->next;
        odd = odd->next;
        even->next = even->next->next;
        even = even->next;
    }

    /* Link the last odd node with first even node and output the resultant list. */
    odd->next = evenhead;
    return head;
}

File: 328.odd-even-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Empty list requires no processing.
        if not head:
            return head

        # Use arrays to store even and odd nodes.
        even = []
        odd = []
        # Use a flag to alternate node addition in even or odd.
        flag = 1
        # Initialize result linked list by adding first odd node.
        newhead = newtail = head
        head = head.next

        while head:
            if flag == 1:
                even.append(head)
            else:
                odd.append(head)

            # Reverse flag to add alternately to even/odd lists.
            flag *= -1
            # Move on to the next node.
            head = head.next

        # Append all odd nodes to result list.
        for node in odd:
            newtail.next = node
            newtail = node

        # Append all even nodes to result list after the odd ones.
        for node in even:
            newtail.next = node
            newtail = node

        # Terminate the list after adding even nodes and return it.
        newtail.next = None
        return newhead
